### PR TITLE
Setup integration tests \o/

### PR DESCRIPTION
**Pull request by vdemeester**
This sets-up the start of real integration tests. This is kinda a big one so let's see what it does 🐼  :

- Updates the build toolchain to build less time binaries (and thus takes a little bit less time). `binary` target is split into `binary`, which build only `dist/traefik`, and `crossbinary`, which use `gox` to build cross-platform binaries.
- Hacks a bit around `circleci` to be able to use the docker daemon it's running on, in order to start external dependencies (`consul`, `marathon`, etc..).
- Sets up integration tests using [libcompose](https://github.com/docker/libcompose), [go-check](https://github.com/go-check/check) and [shakers](https://github.com/vdemeester/shakers). There is multiple suites, and **each suite can start a *compose* project to run external dependencies.**
  
  For now there is the following running suites 🐙 :
    - `SimpleSuite` : which runs really simple cases (like the file does not exists, is invalid, or without any configuration what so ever, traefik starts and listen to port 80).
    - `FileSuite` : which runs tests that uses the `[file]` sections (rules).
    - `DockerSuite` : which runs tests that uses docker as a provider.
    - `ConsulSuite` : which runs tests that uses consul as a provider.
    - `MarathonSuite` : which runs tests that uses marathon as a provider.

  Thoses suites are, in this PR, pretty plain — there will be following PR to add more interesting integration tests — this just sets things up.
- There is a `MAKE_DOCKER_HOST` environment variable switch to tell the container which `DOCKER_HOST` to point to. If not present it will share the `/var/run/docker.sock` via a volume — if present, make sure the host provide is accessible from inside the container.

This looks like :

```bash
$ make test-integration
# […]
docker run  -e "DOCKER_HOST=tcp://172.17.42.1:2375" -it -e OS_ARCH_ARG -e OS_PLATFORM_ARG -e TESTFLAGS -v "/home/ubuntu/traefik/"dist":/go/src/github.com/emilevauge/traefik/"dist"" "traefik-dev:setup-integration-tests" ./script/make.sh generate test-integration
---> Making bundle: generate (in .)

---> Making bundle: test-integration (in .)
PASS: basic_test.go:28: SimpleSuite.TestInvalidConfigShouldFail	0.009s
PASS: basic_test.go:13: SimpleSuite.TestNoOrInexistentConfigShouldFail	0.017s
PASS: basic_test.go:36: SimpleSuite.TestSimpleDefaultConfig	0.102s
PASS: file_test.go:12: FileSuite.TestSimpleConfiguration	0.102s
PASS: docker_test.go:12: DockerSuite.TestSimpleConfiguration	0.102s
PASS: consul_test.go:12: ConsulSuite.TestSimpleConfiguration	0.102s
PASS: marathon_test.go:12: MarathonSuite.TestSimpleConfiguration	0.102s
OK: 7 passed
PASS
ok github.com/emilevauge/traefik/integration 18.595s 
```

There is however few **shortcomings**/**hacks** that will need to be taken care of next/soon :

- As stated in docker/libcompose#32, libcompose is not really *go gettable*, because of the dependencies it has to internal package of `docker/docker`. This will be taken care of at some point in `libcompose` and thus, should be updated here. For now, there is an ugly hack for that (adding a `src/github.com/docker/docker/autogen/dockerversion.go` file somewhere does the trick :sweat_smile:).
- There is an open bug in docker/libcompose#35 (and PR is docker/libcompose#55) that makes somes event to be dropped. Thus, there is some `time.Sleep` piece of code in `BaseSuite` that are here for the short time. As soon as this is fixed, the suites will listen to libcompose events (especially `PROJECT_START_DONE`, `PROJECT_DOWN_DONE`, etc..).
- There is a few duplication is some of the tests ; that should be taken care of, in this PR or in the following one. Let me know :wink: 

I guess that's all for now — waiting for your review @EmileVauge :wink:.

🐸

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,10 +51,6 @@ shell: build
 	$(DOCKER_RUN_TRAEFIK) /bin/bash
 
 image: build
-	if ! [ -a dist/traefik_linux-386 ] ; \
-	then \
-		$(DOCKER_RUN_TRAEFIK) ./script/make.sh generate binary; \
-	fi;
 	docker build -t $(TRAEFIK_IMAGE) .
```
Comment (vdemeester) [inline Makefile]: I'm making the assumption this is mostly always called in `circleci` and/or when the binary already exists.

File: build.Dockerfile [hunk 2]
```diff
@@ -4,10 +4,26 @@ RUN go get github.com/tools/godep
...
+# Which docker version to test on
+ENV DOCKER_VERSION 1.6.2
+
+# Download docker
+RUN set -ex; \
+    curl https://get.docker.com/builds/Linux/x86_64/docker-${DOCKER_VERSION} -o /usr/local/bin/docker-${DOCKER_VERSION}; \
+    chmod +x /usr/local/bin/docker-${DOCKER_VERSION}
+
+# Set the default Docker to be run
+RUN ln -s /usr/local/bin/docker-${DOCKER_VERSION} /usr/local/bin/docker
```
Comment (vdemeester) [inline build.Dockerfile]: This is probably not needed anymore (used it to make my early tests on circleci), tell me if you want me to remove it :sweat_smile:.

File: build.Dockerfile [hunk 3]
```diff
@@ -4,10 +4,26 @@ RUN go get github.com/tools/godep
...
+
 ENV PATH /go/src/github.com/emilevauge/traefik/Godeps/_workspace/bin:$PATH
 
 WORKDIR /go/src/github.com/emilevauge/traefik
 
+# This is a hack (see libcompose#32) - will be removed when libcompose will be fixed
+# (i.e go get able)
+RUN mkdir -p /go/src/github.com/docker/docker/autogen/dockerversion/
+COPY Godeps/_workspace/src/github.com/docker/docker/autogen/dockerversion/dockerversion.go /go/src/github.com/docker/docker/autogen/dockerversion/dockerversion.go
```
Comment (vdemeester) [inline build.Dockerfile]: This is the hack for `docker/docker/autogen` package.

File: circle.yml [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
+    - sudo docker -d -e lxc -s btrfs -H tcp://0.0.0.0:2375:
+        background: true
+    - sleep 5
   environment:
     REPO: $CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME
+    DOCKER_HOST: tcp://172.17.42.1:2375
 
 dependencies:
   pre:
+    - docker version
```
Comment (vdemeester) [inline circle.yml]: Making sure we can talk to docker at this point (fail early, fail often :stuck_out_tongue_closed_eyes:) and also might be a good way to know if circleci updates their docker version :wink:

**Comment by EmileVauge** [comment]
:open_mouth: What a PR ! :metal:

**Comment by thaJeztah** [comment]
> :open_mouth: What a PR ! :metal:

yup, it has the right amount of emoji, must be @vdemeester :smile:

Great work, Vincent!

File: circle.yml [hunk 5]
```diff
@@ -1,11 +1,15 @@
 machine:
-  services:
-    - docker
+  pre:
+    - sudo docker -d -e lxc -s btrfs -H tcp://0.0.0.0:2375:
+        background: true
+    - sleep 5
```
Comment (EmileVauge) [inline circle.yml]: OH! You too!

Comment (vdemeester) [inline circle.yml]: Lol :sweat_smile: yeah.. giving time for docker to start :stuck_out_tongue_closed_eyes:

Comment (EmileVauge) [inline circle.yml]: Maybe we could use ```curl --retry 10 --retry-delay 5 -v http://0.0.0.0:2375/version``` instead :stuck_out_tongue:

Comment (vdemeester) [inline circle.yml]: Lol, that's an idea 🐼

**Comment by EmileVauge** [comment]
This is huge! Thanks a lot @vdemeester :)
Happy testing on Traefik!
LGTM!

**Comment by vdemeester** [comment]
yay :tada:

**Pull request merged**